lib/services/access_request.go

```diff
@@ -1890,6 +2016,25 @@ func UnmarshalAccessRequestAllowedPromotion(data []byte) (*types.AccessRequestAl
 	return &accessListIDs, nil
 }
 
+func getInvalidKubeKindAccessRequestsError(mappedRequestedRolesToAllowedKinds map[string][]string, requestedRoles bool) error {
+	allowedStr := ""
+	for roleName, allowedKinds := range mappedRequestedRolesToAllowedKinds {
+		if len(allowedStr) > 0 {
+			allowedStr = fmt.Sprintf("%s, %s: %v", allowedStr, roleName, allowedKinds)
+		} else {
+			allowedStr = fmt.Sprintf("%s: %v", roleName, allowedKinds)
+		}
+	}
+
+	requestWord := "requestable"
+	if requestedRoles {
+		requestWord = "requested"
+	}
+	return trace.BadParameter(`%s did not allow requesting to some or all of the requested `+
+		`Kubernetes resources. allowed kinds for each %s roles - %v`,
+		InvalidKubernetesKindAccessRequest, requestWord, allowedStr)
+}
+
 // pruneResourceRequestRoles takes an access request and does one of two things:
 //  1. If it is a role request, returns it unchanged.
 //  2. If it is a resource request, all available `search_as_roles` for the user
```

Comment:
While you're here, can you please update the full comment to:
```go
// pruneResourceRequestRoles takes a list of requested resource IDs and 
// a list of candidate roles to request, and returns a "pruned" list of roles.
//
// Candidate roles are *always* pruned when the user is not allowed to
// request the role with all requested resources.
//
// A best-effort attempt is made to prune roles that would not allow
// access to any of the requested resources, this is skipped when any
// resource is in a leaf cluster.
//
// If loginHint is provided, it will attempt to prune the list to a single role.
```